fix test_constant_output

avikchaudhuri opened the pull request:
Summary: Fixes a couple of problems: constants didn't have metadata before creating graph signatures, and graph signatures weren't updated when lifting constants.

Test Plan: fixed test

Differential Revision: D64081786

angelayi commented on torch/export/_trace.py:
```diff
@@ -707,13 +707,18 @@ def _maybe_fixup_gm_and_output_node_meta(old_gm, new_gm):
+        if node.op == "get_attr" and "val" not in node.meta:
+            val = getattr(gm, node.target)
+            if val in constant_attrs and isinstance(val, torch.Tensor):
+                node.meta["val"] = fake_mode.from_tensor(val, static_shapes=True)
```
I'm confused why this is not getting handled [here](https://github.com/pytorch/pytorch/blob/56da78f8faba7a7a9130caddd11c47bad23969b4/torch/_export/passes/lift_constants_pass.py#L229-L237)?

avikchaudhuri replied:
That code is in the lift constant pass, whereas we need to generate a graph signature before that, and the missing meta val causes graph signature generation to crash.

tugsbayasgalan commented on torch/export/_trace.py:
```diff
@@ -707,13 +707,18 @@ def _maybe_fixup_gm_and_output_node_meta(old_gm, new_gm):
                     node.meta["val"] = user_arg
             index += 1
+        if node.op == "get_attr" and "val" not in node.meta:
+            val = getattr(gm, node.target)
```
Would the node.target be in the format of "tensor_constant_0" or sth?

tugsbayasgalan replied:
Maybe assert that node.target is in the format where there is no delimeter in the name. For example, getattr(gm, foo.bar.const) will fail today.

avikchaudhuri replied:
Same code is also here: https://github.com/pytorch/pytorch/blob/56da78f8faba7a7a9130caddd11c47bad23969b4/torch/_export/passes/lift_constants_pass.py#L159

Should both cases use a recursive getattr instead?

tugsbayasgalan replied:
Yeah just to be safe, i think we should. But probably doesn't matter in this case.

tugsbayasgalan commented on torch/export/_trace.py:
```diff
@@ -1531,13 +1536,18 @@ def wrapped_fn(*args):
 
     flat_args = pytree.tree_leaves((fake_args, fake_kwargs))
     index = 0
+    fake_mode = detect_fake_mode(flat_args)
```
This code looks the same as one before. Could we make it a util function?

avikchaudhuri replied:
A lot of the code here is duplicate. Do you want me to pull out more or just this code?

tugsbayasgalan replied:
I think more the better :) But for this PR, just this part is fine.

angelayi commented on torch/_export/passes/lift_constants_pass.py:
```diff
@@ -262,6 +262,10 @@ def lift_constants_pass(
 
+                for spec in graph_signature.output_specs:
+                    if spec.arg.name == node.target:
+                        spec.arg.name = const_placeholder_node.name
```
maybe we can put a mapping of original node -> output spec ahead of time so that we don't have to loop through all the output specs every time we lift a constant?

avikchaudhuri committed:
fix test_constant_output (#137547)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/137547

Fixes a couple of problems: constants didn't have metadata before creating graph signatures, and graph signatures weren't updated when lifting constants.

Test Plan: fixed test

Reviewed By: tugsbayasgalan

Differential Revision: D64081786
